utils/replicator.hh

```diff
@@ -0,0 +1,234 @@
+/// Deletions must be represented as mutations so that they're not lost.
+/// There is no safe garbage-collection implemented yet.
+///
+/// State is not kept here but in the external Container, and its model can extend the Mutation object,
+/// but must be compactible with it. Applying a Mutation to the container should semantically be equivalent
+/// to applying it to a Mutation which represents the state.
+///
+/// Must call and await stop() before destruction.
+template<Mutation mutation_type, Container<mutation_type> container_type>
+class replicator {
+    mutation_type _to_replicate;
+    seastar::condition_variable _replicate_cv;
+    bool _stopping = false;
+    bool _stopped = false;
+    container_type& _container;
+    seastar::future<> _replicate_fiber;
+
+private:
+    seastar::future<> start_replicating() {
+        mutation_type updates;
+        while (true) {
+            co_await _replicate_cv.when([&] {
+                return !updates.empty() || !_to_replicate.empty() || _stopping;
+            });
+
+            if (_stopping) {
+                break;
+            }
+
+            try {
+                if (!updates.empty()) {
+                    // Replication failed
+                    co_await seastar::sleep(std::chrono::seconds(1));
+                    container_type::apply(updates, _to_replicate);
+                } else {
+                    updates = std::move(_to_replicate);
+                }
+                _to_replicate = {};
+                co_await _container.container().invoke_on_others([&](container_type& ctr) {
+                    ctr.apply_locally(updates);
+                });
+                updates = {};
+            } catch (...) {
+                _container.on_replication_failed(std::current_exception());
+            }
+        }
+    }
+
+public:
+    explicit replicator(container_type& ctr)
+        : _container(ctr)
+        , _replicate_fiber(start_replicating())
+    { }
+
+    replicator(replicator&&) = delete; // "this" captured in start_replicating().
+
+    ~replicator() {
+        SCYLLA_ASSERT(_stopped);
+    }
+
+    // The change is applied locally immediately, and to other shards eventually.
+    // If throws, no change is applied anywhere.
+    // Changes are applied in the same order on all shards, so all shards
+    // should eventually converge.
+    template<MutationFragment<container_type, mutation_type> M>
+    void apply_to_all(M m) {
+        container_type::prepare_apply(_to_replicate, m); // ensure [1] below does not fail.
+        _container.apply_locally(m); // Must not throw after this.
+        container_type::apply(_to_replicate, std::move(m)); // [1]
+        _replicate_cv.signal();
+    }
+
+    seastar::future<> stop() {
+        if (_stopping) {
+            seastar::throw_with_backtrace<std::runtime_error>("replicator::stop() called twice");
+        }
+        _stopping = true;
+        _replicate_cv.signal();
+        co_await std::move(_replicate_fiber);
```

Comment:
using std::exchange(_replicate_fiber, make_ready_future()) instead of std::move would be more robust against double close (in production)